kernel/src/process.rs

```diff
@@ -84,6 +69,130 @@ pub unsafe fn load_processes(
+
+    /// Put this process in the fault state. This will trigger the
+    /// `FaultResponse` for this process to occur.
+    unsafe fn fault_state(&self);
+
+    /// Get the name of the process. Used for IPC.
+    fn get_package_name(&self) -> &[u8];
+
+    // memop operations
+
+    /// Change the location of the program break.
+    fn brk(&self, new_break: *const u8) -> Result<*const u8, Error>;
+
+    /// Change the location of the program break and return the previous break
+    /// address.
+    fn sbrk(&self, increment: isize) -> Result<*const u8, Error>;
+
+    /// The start address of allocated RAM for this process.
+    fn mem_start(&self) -> *const u8;
+
+    /// The first address after the end of the allocated RAM for this process.
+    fn mem_end(&self) -> *const u8;
+
+    /// The start address of the flash region allocated for this process.
+    fn flash_start(&self) -> *const u8;
+
+    /// The first address after the end of the flash region allocated for this
+    /// process.
+    fn flash_end(&self) -> *const u8;
+
+    /// The lowest address of the grant region for the process.
+    fn kernel_memory_break(&self) -> *const u8;
+
+    /// How many writeable flash regions defined in the TBF header for this
+    /// process.
+    fn number_writeable_flash_regions(&self) -> usize;
+
+    /// Get the offset from the beginning of flash and the size of the defined
+    /// writeable flash region.
+    fn get_writeable_flash_region(&self, region_index: usize) -> (u32, u32);
+
+    /// Debug function to update the kernel on where the stack starts for this
+    /// process. Processes are not required to call this through the memop
+    /// system call, but it aids in debugging the process.
+    fn update_stack_start_pointer(&self, stack_pointer: *const u8);
+
+    /// Debug function to update the kernel on where the process heap starts.
+    /// Also optional.
+    fn update_heap_start_pointer(&self, heap_pointer: *const u8);
+
+    // additional memop like functions
+
+    /// Check if the buffer address and size is contained within the memory
+    /// owned by this process. This isn't quite the same as the memory allocated
+    /// to the process as this does not include the grant region which is owned
+    /// by the kernel.
+    fn in_app_owned_memory(&self, buf_start_addr: *const u8, size: usize) -> bool;
+
+    /// Get the first address of process's flash that isn't protected by the
+    /// kernel. The protected range of flash contains the TBF header and
+    /// potentially other state the kernel is storing on behalf of the process,
+    /// and cannot be edited by the process.
+    fn flash_non_protected_start(&self) -> *const u8;
+
+    // mpu
+
+    fn setup_mpu(&self, mpu: &mpu::MPU);
+
+    fn add_mpu_region(&self, base: *const u8, size: u32) -> bool;
+
+    // grants
+
+    /// Create new memory in the grant region.
+    unsafe fn alloc(&self, size: usize) -> Option<&mut [u8]>;
+
+    unsafe fn free(&self, _: *mut u8);
```

Comment:
missing doc

also the mpu methods above

Reply:
Well the MPU functions are going to change, so I'm not going to bother documenting those.

Free isn't implemented, so I don't know how to document it.

Reply:
I think these are specific to how the implementation of `ProcessType` decides to handle grants. Unfortunately, how grants are specified is _not_ contained only in the `impl ProcessType` functions. The memory details are also used in `grant.rs` because grants are templated across a `T`, and there is an issue with rust and using that templated value in the `ProcessType` trait. So some of the grant operation logic is handled in process.rs, and some is handled in grant.rs.